contents/blog/secrets-of-posthog-query-performance.md

```diff
@@ -0,0 +1,152 @@
+---
+date: "2022-03-16" # :TODO
+title: "Secrets of PostHog query performance"
+rootPage: /blog
+sidebar: Blog
+showTitle: true
+hideAnchor: true
+featuredImage: ../images/blog/launch-week-teaser.jpeg # :TODO:
+featuredImageType: full
+categories: ["Engineering", "Product updates", "Launch week"]
+author: ["karl-aksel-puulmann"]
+---
+
+We want PostHog to become the first choice for product analytics at any scale. To do that, users should have a smooth experience exploring their product data - including not waiting minutes for queries to load.
+
+In this post, I’m going to break down the why’s and how’s of several major performance improvements we've made in the last few months.
+
+## How does querying work within PostHog?
+
+The free-form querying experience in PostHog allows you to ask questions about your Trends, Funnels, Retention, and Cohorts with complicated filtering to top it off.
+
+These queries are processed by [ClickHouse](https://clickhouse.com/), where event, user, and group data is stored in a raw format without any preaggregation. ClickHouse excels at analyzing billions of events in seconds even with complex filtering.
+
+However, as awesome as ClickHouse is, nothing is without sharp edges and trade-offs.
+
+## Speeding up property filtering (up to) 25x
+
+PostHog allows users to send and analyze arbitrary number of event and user properties with their data. We store this data as JSON-encoded strings in our tables.
+
+One of the first issues we saw after moving to ClickHouse was that, for our largest users, filtering by properties was slow.
+
+[Benchmarking these queries using flamegraphs](https://github.com/Slach/clickhouse-flamegraph) showed that the slowness came from two things: reading JSON properties from disk and (to a lesser extent) parsing it during query-time.
+
```

Comment:
What about adding the flamegraphs here?

Reply:
It's a good question - I think they're hard to interpret without additional context so going to leave them out for now.